src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.ValueList.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Text.Json
+{
+    internal sealed partial class JsonPropertyDictionary<T>
+    {
+#if DEBUG
+        // CreateEditableValueList should be called at most once by JsonTypeInfo
+        private ValueList? _editableValueList;
+#endif
+
+        public ValueList CreateValueList(Func<T, string>? getKey)
+        {
+            ValueList ret = new ValueList(this, getKey);
+#if DEBUG
+            Debug.Assert(_editableValueList == null, "More than one ValueList created");
+            return _editableValueList ??= ret;
+#else
+            return ret;
+#endif
+        }
+
+        internal sealed class ValueList : IList<T>
+        {
+            private readonly JsonPropertyDictionary<T> _parent;
+            private Func<T, string>? _getKey;
```

Comment:
QQ, why does key extraction need to be extracted behind a delegate? In main it looks like we're storing it as a `List<KeyValuePair<string, T?>>`

Reply:
it's because changing JsonPropertyInfo.Name will also change the key and we do not want to keep track of list JsonPropertyInfo is inserted into - instead we do it once in the end of editing

Reply:
Is the `T` being used with generic parameters other than `JsonPropertyInfo`? If not we could just make it non-generic.

Reply:
yes, it's used in the fast path but overall I'm seeing 18 refs